### PR TITLE
Islot changes

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
Introduce a new data structure i.e. islot_info from icap. 
XCLBIN related information is separated from icap structure and added to the new islot_info data structure. 

This is required for multislot support. As we need to cache multiple xclbin information. 
Right now there is no functionality changes. Only data structure changes. 

Also make sure that changes are concentrate within icap.c file. Hence used slot-0 as default slot. 
All the other APIs which has a referencs of some other file remain unchanged and used slot 0 as default slot.

Basic xbutil validate is working fine with this changes

### CODE REVIEW

**Review by mamin506 — CHANGES_REQUESTED** [review]
I'm a little bit worry about the locking. I saw that at a lot of places, icap_lock is simply replaced by islot_lock. The critical section of this two locks are different. It looks like the new code changes the original locking design. I'm wondering if this is safe.

I'm not very familiar with the icap locking design. Maybe my concern is not true.
Max has more knowledge on icap than I do. Let's see what he comments on this.

**Review by maxzhen — CHANGES_REQUESTED** [review]
The per-islot (BTW, why 'i'slot? why not just slot?) lock does not work well. There are cases where you need a lock across all slots. E.g.,
static void platform_reset_axi_gate(struct platform_device *pdev)
{
	struct icap *icap = platform_get_drvdata(pdev);
	uint32_t slot_id = DEFAULT_SLOT_ID; // Default Slot

	/* Can only be done from mgmt pf. */
	if (!ICAP_PRIVILEGED(icap))
		return;

	mutex_lock(&icap->icap_lock);
	if (!icap_bitstream_in_use(icap, slot_id)) {
		(void) icap_freeze_axi_gate(platform_get_drvdata(pdev));
		(void) icap_free_axi_gate(platform_get_drvdata(pdev));
	}
	mutex_unlock(&icap->icap_lock);
}
The axi gate resetting may impact ALL slots, but you just check in-use state of one of them to determine if axi-gate can be reset or not. This does not look right. There maybe other cases...

Anyway, why do you need a per-slot lock? Can we just have one big lock for all slots so that locking schema does not need to change?

**Comment by saifuddin-xilinx** [comment]
retest this please

**Comment by saifuddin-xilinx** [comment]
> The per-islot (BTW, why 'i'slot? why not just slot?) lock does not work well. There are cases where you need a lock across all slots. E.g., static void platform_reset_axi_gate(struct platform_device *pdev) { struct icap *icap = platform_get_drvdata(pdev); uint32_t slot_id = DEFAULT_SLOT_ID; // Default Slot
> 
> ```
> /* Can only be done from mgmt pf. */
> if (!ICAP_PRIVILEGED(icap))
> 	return;
> 
> mutex_lock(&icap->icap_lock);
> if (!icap_bitstream_in_use(icap, slot_id)) {
> 	(void) icap_freeze_axi_gate(platform_get_drvdata(pdev));
> 	(void) icap_free_axi_gate(platform_get_drvdata(pdev));
> }
> mutex_unlock(&icap->icap_lock);
> ```
> 
> } The axi gate resetting may impact ALL slots, but you just check in-use state of one of them to determine if axi-gate can be reset or not. This does not look right. There maybe other cases...
> 
> Anyway, why do you need a per-slot lock? Can we just have one big lock for all slots so that locking schema does not need to change?

This seems to a good idea. I am agree to having a single lock. It will reduce the chance to introduce new BUG.
I will make the necessary changes.  Thanks.

**Comment by salindac** [comment]
retest this please

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 1]
```diff
@@ -1357,12 +1467,14 @@ static uint16_t icap_get_memidx(struct mem_topology *mem_topo, enum IP_TYPE ecc_
...
-static int icap_create_subdev_debugip(struct platform_device *pdev)
+static int icap_create_subdev_debugip(struct platform_device *pdev,
+				      uint32_t slot_id)
 {
 	struct icap *icap = platform_get_drvdata(pdev);
 	int err = 0, i = 0;
 	xdev_handle_t xdev = xocl_get_xdev(pdev);
-	struct debug_ip_layout *debug_ip_layout = icap->debug_layout;
+	struct debug_ip_layout *debug_ip_layout =
+		icap->slot_info[slot_id]->debug_layout;
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:1477]: Can slot_info[slot_id] be NULL in this function? Just like many other functions.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 2]
```diff
@@ -1507,13 +1619,14 @@ static int icap_create_subdev_debugip(struct platform_device *pdev)
...
+static int icap_create_subdev_ip_layout(struct platform_device *pdev,
+					uint32_t slot_id)
 {
 	struct icap *icap = platform_get_drvdata(pdev);
 	int err = 0, i = 0;
 	xdev_handle_t xdev = xocl_get_xdev(pdev);
-	struct ip_layout *ip_layout = icap->ip_layout;
-	struct mem_topology *mem_topo = icap->mem_topo;
+	struct ip_layout *ip_layout = icap->slot_info[slot_id]->ip_layout;
+	struct mem_topology *mem_topo = icap->slot_info[slot_id]->mem_topo;
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:1629]: Ditto.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 3]
```diff
@@ -1636,13 +1749,14 @@ static int icap_create_subdev_ip_layout(struct platform_device *pdev)
...
+static int icap_create_post_download_subdevs(struct platform_device *pdev,
+					 struct axlf *xclbin, uint32_t slot_id)
 {
 	struct icap *icap = platform_get_drvdata(pdev);
 	int err = 0, i = 0;
 	xdev_handle_t xdev = xocl_get_xdev(pdev);
-	struct ip_layout *ip_layout = icap->ip_layout;
-	struct mem_topology *mem_topo = icap->mem_topo;
+	struct ip_layout *ip_layout = icap->slot_info[slot_id]->ip_layout;
+	struct mem_topology *mem_topo = icap->slot_info[slot_id]->mem_topo;
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:1759]: Ditto.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 4]
```diff
@@ -208,88 +218,129 @@ static inline void reg_wr(void __iomem *reg, u32 val)
...
 
-	ret = wait_event_interruptible(icap->reader_wq, icap->reader_ref == 0);
+	ret = wait_event_interruptible(islot->reader_wq,
+				       islot->reader_ref == 0);
 
 	if (ret)
 		goto done;
 
-	BUG_ON(icap->reader_ref != 0);
+	BUG_ON(islot->reader_ref != 0);
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:277]: We should move "islot = icap->slot_info[slot_id];" into the critical section. The icap_lock is also protecting slot_info array.

**Review by mamin506 — CHANGES_REQUESTED** [review]
@saifuddin-xilinx , the icap_lock issue is gone. Can you address my comments? So that I can move forward to approve this.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 5]
```diff
@@ -135,17 +138,9 @@ struct icap_bitstream_user {
...
-	struct icap_generic_state *icap_state;
-	struct icap_config_engine *icap_config_engine;
-	unsigned int		idcode;
-	bool			icap_axi_gate_frozen;
-
-	xuid_t			icap_bitstream_uuid;
-	int			icap_bitstream_ref;
+struct islot_info {
+	uint32_t		slot_idx;
+	uint32_t		pr_idx;
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:143]: This is added for PL slot. For single slot PL case, this doesn't have any significance.  We can remove it for now as we are strict to single PL slot and multiple PS slot.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 1 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:1477]: Fixed. Thanks

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 3 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:1759]: Fixed.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 2 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:1629]: Fixed

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 4 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:277]: I am not sure, As we fallback to icap lock, do we need to protect the slot array or not. 
This is the same behavior as of today without slot changes. Am i missing something ?

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 6]
```diff
@@ -208,88 +216,129 @@ static inline void reg_wr(void __iomem *reg, u32 val)
...
-	int *num_urpdev, struct xocl_subdev **urpdevs);
+	int *num_urpdev, struct xocl_subdev **urpdevs, uint32_t slot_id);
+
+static struct islot_info *
+icap_slot_init(struct icap *icap, uint32_t slot_id)
+{
+	struct islot_info *islot = icap->slot_info[slot_id];
+	
+	if (islot) {
+		vfree(islot);
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:234]: you freed the pointer but left a dangling one in icap->slot_info[slot_id]?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 7]
```diff
@@ -208,88 +216,129 @@ static inline void reg_wr(void __iomem *reg, u32 val)
...
+	islot = vzalloc(sizeof(struct islot_info));
+	if (!islot) {
+		ICAP_ERR(icap, "Memory allocation failure");
+		return NULL;
+	}
+	
+	init_waitqueue_head(&islot->reader_wq);
+	islot->slot_idx = slot_id;
+
+	return islot;
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:246]: why not assign it back to icap->slot_info[slot_id]? if you just want to alloc a struct islot_info in this function, then don't touch the global icap->slot_info[slot_id], or you should hold the lock, free the original one, and add new one to it, release the lock.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 8]
```diff
@@ -208,88 +216,129 @@ static inline void reg_wr(void __iomem *reg, u32 val)
...
+}
 
-static int icap_xclbin_wr_lock(struct icap *icap)
+static int icap_xclbin_wr_lock(struct icap *icap, uint32_t slot_id)
 {
+	struct islot_info *islot = icap->slot_info[slot_id];
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:251]: don't you need to grab a lock before accessing this global array? what if islot is freed by another thread right after  you cached the pointer in islot here?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 4 above)
Comment (maxzhen) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:275]: the icap lock was used to protect icap data structure, now you added slot array to it, why it's not protected by it?

**Review by maxzhen — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 9]
```diff
@@ -2551,12 +2712,24 @@ static int icap_download_bitstream_axlf(struct platform_device *pdev,
...
 
-	err = icap_xclbin_wr_lock(icap);
-	if (err)
+	/* This is the first entry for slot in icap. 
+	 * Hence allocate required memory here */
+	islot = icap_slot_init(pdev, slot_id);
+	if (islot == NULL)
+		return -ENOMEM;
+
+	icap->slot_info[slot_id] = islot;
```
[the remark this answers is not in the document]
Comment (mamin506) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:2719]: Any comment on this one? @saifuddin-xilinx

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 10]
```diff
@@ -208,88 +219,131 @@ static inline void reg_wr(void __iomem *reg, u32 val)
...
-static void icap_xclbin_wr_unlock(struct icap *icap)
+
+static void icap_xclbin_wr_unlock(struct icap *icap, uint32_t slot_id)
 {
+	struct islot_info	*islot = icap->slot_info[slot_id];
 	pid_t pid = pid_nr(task_tgid(current));
 
-	BUG_ON(icap->busy != (u64)pid);
+	if (!islot)
+		return;
```
[the remark this answers is not in the document]
Comment (mamin506) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:288]: Any comment on this one? @saifuddin-xilinx

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 11]
```diff
@@ -411,10 +465,15 @@ static unsigned short icap_cached_ocl_frequency(const struct icap *icap, int idx
...
 
-static bool icap_bitstream_in_use(struct icap *icap)
+static bool icap_bitstream_in_use(struct icap *icap, uint32_t slot_id)
 {
-	BUG_ON(icap->icap_bitstream_ref < 0);
-	return icap->icap_bitstream_ref != 0;
+	struct islot_info *islot = icap->slot_info[slot_id];
+
+	if (!islot)
+		return false;
```
[the remark this answers is not in the document]
Comment (mamin506) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:468]: Any comment on this one? @saifuddin-xilinx

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mamin506) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: So, what are PL slot and PS slot? Do we have any confluence pages?

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by saifuddin-xilinx** [comment]
retest this please

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: Updated. Please check.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: Fixed. Thanks

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: Removed.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 4 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: Done. Please check. Thanks

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 6 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:234]: Fixed.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 7 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: Fixed. Thanks

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 8 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: Added lock.

**Review by mamin506 — CHANGES_REQUESTED** [review]
One minor change.
One lock design question needs to be address.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 12]
```diff
@@ -3151,31 +3322,40 @@ static ssize_t clock_freqs_show(struct device *dev,
...
-	mutex_lock(&icap->icap_lock);
-	for (i = 0; i < ICAP_MAX_NUM_CLOCKS; i++) {
-		freq = icap_get_ocl_frequency(icap, i);
+		err = icap_xclbin_rd_lock(icap, st);
+		if (err)
+			return cnt;
 
-		if (!uuid_is_null(&icap->icap_bitstream_uuid)) {
-			freq_counter = icap_get_clock_frequency_counter_khz(icap, i);
+		mutex_lock(&icap->icap_lock);
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:3341]: Somehow this PR code is not clear. Please check the actual code block attached below. 
` 
       for (st = 0; st < MAX_SLOT_SUPPORT; st++) {
                islot = icap->slot_info[st];
                if (islot == NULL)
                        continue;

                err = icap_xclbin_rd_lock(icap, st);
                if (err)
                        return cnt;

                mutex_lock(&icap->icap_lock);
                for (i = 0; i < ICAP_MAX_NUM_CLOCKS; i++) {
                        freq = icap_get_ocl_frequency(icap, i);

                        if (!uuid_is_null(&islot->icap_bitstream_uuid)) {
                                freq_counter = icap_get_clock_frequency_counter_khz(icap, i, st);

                                request_in_khz = freq*1000;
                                tolerance = freq*50;

                        if (abs(freq_counter-request_in_khz) > tolerance)
                                ICAP_INFO(icap, "Frequency mismatch, Should be %u khz, Now is %ukhz", request_in_khz, freq_counter);
                        cnt += sprintf(buf + cnt, "%d\n", DIV_ROUND_CLOSEST(freq_counter, 1000));
                } else
                                cnt += sprintf(buf + cnt, "%d\n", freq);
                }

                mutex_unlock(&icap->icap_lock);
                icap_xclbin_rd_unlock(icap, st);
`

**Review by mamin506 — APPROVED** [review]
For current DEFAULT slot code, this change is okay.

**Review by maxzhen — APPROVED** [review]

**Review by chvamshi-xilinx — APPROVED** [review]

**Comment by chvamshi-xilinx** [comment]
@saifuddin-xilinx please rebase your change

**Pull request merged**